### PR TITLE
Convert to package, add support for external process management

### PR DESCRIPTION
**Pull request by faisal-memon**
Full list of changes:

- Move spiffe-helper.go to pkg/ directory and make it part of spiffe package rather than main
- Move main.go and config.go to cmd/ directory
- Add ExternalProcess flag to signify process will be managed externally
- Add ProcessReloadFunc to specify external function that will be called to reload process when new cert is received.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,15 +3,15 @@ export GO111MODULE=on
 .PHONY: all build test clean distclean
 
 build:
-	go build
+	go build -o spiffe-helper cmd/spiffe-helper/*.go
```
Comment (azdagron) [inline Makefile]: Using go build in this way (building specific .go files) is not recommended. You should instead build the package:

```suggestion
	go build -o spiffe-helper ./cmd/spiffe-helper
```

File: cmd/spiffe-helper/config_test.go [hunk 2]
```diff
@@ -6,7 +6,7 @@ import (
 )
 
 func TestParseConfig(t *testing.T) {
-	c, err := ParseConfig("./test/fixture/config/helper.conf")
+	c, err := ParseConfig("../..//test/fixture/config/helper.conf")
```
Comment (azdagron) [inline cmd/spiffe-helper/config_test.go]:
```suggestion
	c, err := ParseConfig("../../test/fixture/config/helper.conf")
```

File: pkg/helper/helper.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
+package spiffe
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/helper/helper.go]: I personally would prefer `sidecar` (and therefore `github.com/spiffe/spiffe-helper/pkg/sidecar`). `helper` feels way too generic as a package name.

File: go.sum [hunk 4]
```diff
@@ -205,6 +205,8 @@ github.com/spiffe/go-spiffe v0.0.0-20170907221946-2bb3101d62b4/go.mod h1:q0X9/v7
 github.com/spiffe/go-spiffe v0.0.0-20190717182101-d8657cb50cae/go.mod h1:HyNeJnVYkDyQgB2qcSPxVYkAA2F3lQu51bDxNpFcKxY=
 github.com/spiffe/go-spiffe v0.0.0-20190922191205-018e7197ed1c h1:wpwh25WjvKF8/+N+wMy1u9nMiOXfw5sqpmL5ZSAFIWU=
 github.com/spiffe/go-spiffe v0.0.0-20190922191205-018e7197ed1c/go.mod h1:HyNeJnVYkDyQgB2qcSPxVYkAA2F3lQu51bDxNpFcKxY=
+github.com/spiffe/spiffe-helper v0.0.0-20191008190511-e11122b96bb9 h1:4Ze5T6YLet2KlPoGXCS0yshA/zv/EkJmirbKUL1PQ08=
```
Comment (azdagron) [inline go.sum]: This change should be reverted once you rename the module back to github.com/spiffe/spiffe-helper

File: cmd/spiffe-helper/config.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package main
```
Comment (azdagron) [inline cmd/spiffe-helper/config.go]: file could benefit from a run through `gofmt`

File: pkg/helper/helper_test.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package main
+package spiffe
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/helper/helper_test.go]: Problem i have with both of those suggestions is that neither ties this package to being part of spiffe.

File: go.sum (hunk 4 above)
Comment (faisal-memon) [reply go.sum]: done

File: cmd/spiffe-helper/config.go (hunk 5 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: done

File: cmd/spiffe-helper/config.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	dat, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse HCL
+	if err := hcl.Decode(helperConfig, string(dat)); err != nil {
+		return nil, fmt.Errorf("unable to decode configuration: %v", err)
+	}
+	if err != nil {
```
Comment (amartinezfayo) [inline cmd/spiffe-helper/config.go]: This check can be removed.

Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: I had to change back to hcl.Parse() and hcl.DecoeObject() because hcl.Decode does not remove the `hcl:...` tags from the struct if the field is not populated in the config file.

File: pkg/helper/helper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/helper/helper.go]: Changed to sidecar.

File: pkg/helper/helper.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
+	"github.com/spiffe/spire/api/workload"
+	"golang.org/x/sys/unix"
+)
+
+// HelperConfig contains config variables when creating a SPIFFE Helper.
+type HelperConfig struct {
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/helper/helper.go]: done

File: cmd/spiffe-helper/config.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	"io/ioutil"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spiffe-helper/pkg/sidecar"
+)
+
+// ParseConfig parses the given HCL file into a SidecarConfig struct
+func ParseConfig(file string) (Config *sidecar.Config, err error) {
+	sidecarConfig := &sidecar.Config{
+		AgentAddress:       `hcl:"agentAddress"`,
```
Comment (azdagron) [inline cmd/spiffe-helper/config.go]: the fields are being initialized to hcl struct tags? I think we probably just want the following:
```suggestion
		sidecarConfig := new(sidecar.Config)
```
Those struct tags belong on the sidecar.Config fields...

File: cmd/spiffe-helper/config.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	// Read HCL file
+	dat, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	hclText := string(dat)
+
+	// Parse HCL
+        hclParseTree, err := hcl.Parse(hclText)
```
Comment (azdagron) [inline cmd/spiffe-helper/config.go]: gofmt

File: cmd/spiffe-helper/config.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        if err != nil {
+                return nil, err
+        }
+
+        if err := hcl.DecodeObject(&sidecarConfig, hclParseTree); err != nil {
```
Comment (azdagron) [inline cmd/spiffe-helper/config.go]: sidecarConfig is already a pointer. You don't need to take the address and can just pass it directly. Also the parse and then decodeobject can be simplified to the following:
```suggestion
        if err := hcl.Decode(&sidecarConfig, string(dat)); err != nil {
```

File: go.mod [hunk 12]
```diff
@@ -1,20 +1,14 @@
 module github.com/spiffe/spiffe-helper
 
+replace github.com/spiffe/spiffe-helper => ./
```
Comment (azdagron) [inline go.mod]: This replace statement should be removed.

File: pkg/sidecar/sidecar.go [hunk 13]
```diff
@@ -17,59 +17,96 @@ import (
...
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
-	proto "github.com/spiffe/spire/proto/api/workload"
+	"golang.org/x/sys/unix"
 )
 
-// sidecar is the component that consumes the Workload API and renews certs
+// Config contains config variables when creating a SPIFFE Sidecar.
+type Config struct {
+	AgentAddress       string
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: hcl struct tags that were left in ParseConfig when the unnamed struct was extracted should be restored here

File: pkg/sidecar/sidecar.go [hunk 14]
```diff
@@ -17,59 +17,96 @@ import (
...
 	go func() {
-		err := s.workloadAPIClient.Start()
-		if err != nil {
-			log.Println(err.Error())
-			errorChan <- err
+		clk := clock.New()
+		delay := delayMin
+		for {
+			err := s.workloadAPIClient.Start()
+			if err != nil {
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: What problem was this change trying to solve? As configured, Start() won't return until Stop has been called. The client already takes care of backoff and executing retries internally...

File: pkg/sidecar/sidecar.go [hunk 15]
```diff
@@ -17,59 +17,96 @@ import (
...
+		certReadyChan:     make(chan struct{}, 1),
 	}, nil
 }
 
 // RunDaemon starts the main loop
 // Starts the workload API client to listen for new SVID updates
 // When a new SVID is received on the updateChan, the SVID certificates
 // are stored in disk and a restart signal is sent to the proxy's process
-func (s *sidecar) RunDaemon(ctx context.Context) error {
+func (s *Sidecar) RunDaemon(ctx context.Context) {
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: From a library perspective, I think it makes sense for RunDaemon to continue returning an error in case something catatrophic (i.e. non-temporary) does not allow it do its job.

File: pkg/sidecar/sidecar.go [hunk 16]
```diff
@@ -17,59 +17,96 @@ import (
...
 	// Create channel for interrupt signal
 	interrupt := make(chan os.Signal, 1)
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: TERM and INT should probably not be handled from inside the library since each application would have its own requirements on what causes `RunDaemon` to return. I'd suggest moving the signal handling back out into `main` and change it to just cancel the context that is passed into `RunDaemon`.

File: pkg/sidecar/sidecar.go [hunk 17]
```diff
@@ -17,59 +17,96 @@ import (
...
+	AgentAddress       string
+	Cmd                string
+	CmdArgs            string
+	CertDir            string
+	SvidFileName       string
+	SvidKeyFileName    string
+	SvidBundleFileName string
+	RenewSignal        string
+	Timeout            string
+	ExternalProcess    bool
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: What's the use case for `ExternalProcess` and `ProcessReloadFunc`?

File: pkg/sidecar/sidecar.go [hunk 18]
```diff
@@ -17,59 +17,96 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/andres-erbsen/clock"
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: The spiffe-helper should start consuming the newer go-spiffe/workload package instead of depending on SPIRE but that is outside of the scope of this change.

File: pkg/sidecar/sidecar.go [hunk 19]
```diff
@@ -101,9 +139,19 @@ func updateCertificates(s *sidecar, svidResponse *proto.X509SVIDResponse) {
...
 		log.Println(err.Error())
 	}
+
+	select {
+	case s.certReadyChan <- struct{}{}:
+	default:
+	}
+}
+
+// CertReadyChan returns a channel to know when the trust bundle is ready
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: nit: `when the certificates are ready`

File: cmd/spiffe-helper/config.go (hunk 9 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: I moved the hcl stuff out here because we use this as a package and don't load config from a file. I thought it made more sense to have the struct tags closer to where the config is read from a file. Happy to put it back where it was if you think thats a better approach.

Comment (azdagron) [reply cmd/spiffe-helper/config.go]: I see. In that case, we probably want to duplicate the struct into the `main` package, attach the `hcl` tags and initialize `sidecar.Config` with that struct. It does make it so that newly added config needs to be threaded through one more spot, but that makes supporting a new sidecar configuration from `spiffe-helper` more explicit (since you explicitly have to add a new field to the "hcl" struct, update documentation, etc.)

In either case, these field assignments aren't doing anything useful other than initializing the struct field values with garbage.

File: cmd/spiffe-helper/config.go (hunk 10 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: done

File: pkg/sidecar/sidecar.go (hunk 18 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: Thanks, Ill look into that after this is merged.

File: go.mod (hunk 12 above)
Comment (faisal-memon) [reply go.mod]: done

File: pkg/sidecar/sidecar.go (hunk 14 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: I didnt change this section of the code so Im not sure why its showing up here. This code exists in master.

Comment (azdagron) [reply pkg/sidecar/sidecar.go]: Oh weird. I think your chain of commits has some of the other commits that were merged into master in there. I'm not sure if they were cherry-picked or something but don't seem have been introduced into your branch via a merge commit so they show up as explicit commits in your PR, which is why github is showing it the way that it is.

We're pretty close on this and there are some merge conflicts with master. Maybe the right thing to do is to rebase this on top of master, fix up the conflicts and squash it all down to a single commit? Then we're not trying to review code that you didn't introduce...

File: pkg/sidecar/sidecar.go (hunk 15 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: I made this change so that RunDaemon can be started as a go routine. We need it to return to so that we can do other things, like monitor for metrics, etc. Perhaps there is a better way to structure it.

Comment (azdagron) [reply pkg/sidecar/sidecar.go]: One way of structuring it would be like so: ```func DoSomeThingsUntilOneOfThemFailsOrContextIsCanceled(ctx context.Context) {
	var wg sync.WaitGroup
	defer wg.Wait()

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	// make the channel buffer size enough to hold an error from all the
	// goroutines you care about (two goroutines in this example).
	errsCh := make(chan error, 2)

	wg.Wait(1)
	go func() {
		defer wg.Done()
		errsCh <- sidecar.RunDaemon(ctx)
	}()

	// maybe kick off other goroutines to do other things
	wg.Wait(1)
	go func() {
		defer wg.Done()
		errsCh <- DoSomethingElse(ctx)
	}()

	select {
	case err := errsCh:
		// returning here will cause the defers above to run, canceling the
		// context and wait for the rest of the goroutines to finish.
		return err
	case <-ctx.Done():
		// The context has been canceled by a higher layer, then the functions
		// running in the above goroutines (which should be written to honor
		// context cancellation) will return. The defer's above will cancel the
		// context (noop since it has already been canceled) and then wait
		// for the goroutines to finish.
		return ctx.Err()
	}
}```

Comment (azdagron) [reply pkg/sidecar/sidecar.go]: The above is of course only good if you aren't expecting any of the goroutines to finish early (you'd have to make some adjustments)

Comment (azdagron) [reply pkg/sidecar/sidecar.go]: I think the key point is that even if you kick something off in a goroutine you probably still care when that thing is done (failed or otherwise)

Comment (azdagron) [reply pkg/sidecar/sidecar.go]: (fixed the above to actually compile, but I haven't run it)

File: cmd/spiffe-helper/config.go (hunk 9 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: I moved it back to the sidecar.go and that works fine. If the tags aren't used they are ignored.

File: cmd/spiffe-helper/config.go (hunk 11 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config.go]: done.

File: pkg/sidecar/sidecar.go (hunk 13 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: done

File: pkg/sidecar/sidecar.go (hunk 19 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: fixed

File: pkg/sidecar/sidecar.go (hunk 14 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: This is done.

File: pkg/sidecar/sidecar.go (hunk 15 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: Thanks for that, it was very helpful. Instead of using a errorCh i had to do err := RunDaemon() because the return type of RunDaemon is error.

I added the error return back to RunDaemon.

File: pkg/sidecar/sidecar.go (hunk 18 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: Added a task in our internal jira for this.

File: pkg/sidecar/sidecar.go (hunk 16 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: Done.

File: pkg/sidecar/sidecar.go [hunk 20]
```diff
@@ -10,25 +10,40 @@ import (
...
+	AgentAddress       string `hcl:"agentAddress"`
+	Cmd                string `hcl:"cmd"`
+	CmdArgs            string `hcl:"cmdArgs"`
+	CertDir            string `hcl:"certDir"`
+	SvidFileName       string `hcl:"svidFileName"`
+	SvidKeyFileName    string `hcl:"svidKeyFileName"`
+	SvidBundleFileName string `hcl:"svidBundleFileName"`
+	RenewSignal        string `hcl:"renewSignal"`
+	Timeout            string `hcl:"timeout"`
+	ExternalProcess    bool
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: Perhaps since ExternalProcess can't be set without ProcessReloadFunc (or the library will panic) and ProcessReloadFunc is meaningless without External process being set, maybe we should just get rid of the ExternalProcess bool? and maybe rename ProcessReloadFunc to `ReloadExternalProcess`?

File: pkg/sidecar/sidecar.go [hunk 21]
```diff
@@ -135,45 +157,66 @@ func newWorkloadAPIClient(agentAddress string, timeout time.Duration) workload.X
...
+			if err != nil {
+				return fmt.Errorf("error signaling process with signal: %v\n%v", sig, err)
+			}
 		}
 
-		err = s.process.Signal(sig)
-		if err != nil {
-			return fmt.Errorf("error signaling process with signal: %v\n%v", sig, err)
+	case true:
+		if atomic.LoadInt32(&s.processRunning) == 1 {
```
Comment (azdagron) [inline pkg/sidecar/sidecar.go]: It seems a little weird for an outside caller to have to educate the library about whether or not an externally managed process is running (via ProcessStarted) just so the library can know whether or not it should invoke the caller supplied callback.

I think things would be much simpler (and require less outside coordination with the library) if the invocation of ProcessReloadFunc didn't care about `s.processRunning` (and consequently `ProcessStarted` can be removed. The outside caller will already need to be tracking the lifetime of the external process and so can easily decide to ignore a ProcessReloadFunc callback if the process isn't running.

File: pkg/sidecar/sidecar.go (hunk 20 above)
Comment (azdagron) [reply pkg/sidecar/sidecar.go]: (where we determine whether or not to manage the external process by whether or not `ReloadExternalProcess` is set)

Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: done

File: pkg/sidecar/sidecar.go (hunk 21 above)
Comment (faisal-memon) [reply pkg/sidecar/sidecar.go]: done
